llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp

```diff
@@ -588,10 +588,13 @@ void VPlanTransforms::handleEarlyExits(VPlan &Plan,
         handleUncountableEarlyExit(cast<VPBasicBlock>(Pred), EB, Plan,
                                    cast<VPBasicBlock>(HeaderVPB), LatchVPBB);
         HandledUncountableEarlyExit = true;
-      } else {
+      }
+
+      if (!HasUncountableEarlyExit ||
+          Plan.shouldEarlyExitContinueInScalarLoop())
```

Comment:
Can't you just pass this flag in as an argument to handleEarlyExits, similar to `HasUncountableEarlyExit`? If you follow @sdesmalen-arm's suggestion for an enum, then you could just pass in a single enum here.